pkg/providers/cloudstack/validator.go

```diff
@@ -28,104 +29,136 @@ var restrictedUserCustomDetails = [...]string{
+
+		zoneId, err := v.cmk.ValidateZonePresent(ctx, az.CredentialsRef, az.CloudStackAvailabilityZone.Zone)
 		if err != nil {
-			return fmt.Errorf("checking network %v", err)
+			return err
+		}
+		az.CloudStackAvailabilityZone.Zone.Id = zoneId
+		if len(az.CloudStackAvailabilityZone.Zone.Network.Id) == 0 && len(az.CloudStackAvailabilityZone.Zone.Network.Name) == 0 {
+			return fmt.Errorf("zone network is not set or is empty")
+		}
+		if err := v.cmk.ValidateNetworkPresent(ctx, az.CredentialsRef, az.DomainId, az.CloudStackAvailabilityZone.Zone.Network, zoneId, az.Account, true); err != nil {
+			return err
 		}
 	}
 
 	logger.MarkPass("Datacenter validated")
 	return nil
 }
 
-func (v *Validator) validateDomainAndAccount(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) error {
-	if (datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account == "") ||
-		(datacenterConfig.Spec.Domain == "" && datacenterConfig.Spec.Account != "") {
-		return fmt.Errorf("both domain and account must be specified or none of them must be specified")
+func generateLocalAvailabilityZones(ctx context.Context, datacenterConfig *anywherev1.CloudStackDatacenterConfig) ([]localAvailabilityZone, error) {
+	localAvailabilityZones := []localAvailabilityZone{}
+
+	if datacenterConfig == nil {
+		return nil, errors.New("CloudStack Datacenter Config is null")
 	}
 
-	if datacenterConfig.Spec.Domain != "" && datacenterConfig.Spec.Account != "" {
-		domain, errDomain := v.cmk.ValidateDomainPresent(ctx, datacenterConfig.Spec.Domain)
-		if errDomain != nil {
-			return fmt.Errorf("checking domain: %v", errDomain)
+	if len(datacenterConfig.Spec.Domain) > 0 {
+		for index, zone := range datacenterConfig.Spec.Zones {
+			availabilityZone := localAvailabilityZone{
+				CloudStackAvailabilityZone: &anywherev1.CloudStackAvailabilityZone{
+					Name:                  fmt.Sprintf("availability-zone-%d", index),
+					CredentialsRef:        decoder.CloudStackGlobalAZ,
+					Domain:                datacenterConfig.Spec.Domain,
+					Account:               datacenterConfig.Spec.Account,
+					ManagementApiEndpoint: datacenterConfig.Spec.ManagementApiEndpoint,
+					Zone:                  zone,
+				},
+			}
+			localAvailabilityZones = append(localAvailabilityZones, availabilityZone)
 		}
-
-		errAccount := v.cmk.ValidateAccountPresent(ctx, datacenterConfig.Spec.Account, domain.Id)
-		if errAccount != nil {
-			return fmt.Errorf("checking account: %v", errAccount)
+	}
+	for _, az := range datacenterConfig.Spec.AvailabilityZones {
+		availabilityZone := localAvailabilityZone{
+			CloudStackAvailabilityZone: &az,
 		}
+		localAvailabilityZones = append(localAvailabilityZones, availabilityZone)
+	}
 
-		domainId = domain.Id
+	if len(localAvailabilityZones) <= 0 {
+		return nil, fmt.Errorf("CloudStackDatacenterConfig domain or localAvailabilityZones is not set or is empty")
```

Comment:
```suggestion
		return nil, fmt.Errorf("CloudStackDatacenterConfig domain or AvailabilityZones is not set or is empty")
```
users wouldn't be able to set localAvailabilityZones right?

Reply:
Good find.